Add builder for Backoff

kth496 opened the pull request:
Motivation:

- #5483

Modifications:

- Add `Builder` class for various backoff implementations.

Result:

- Closes #5483
- Will be able to use the builder class to create instances of BackOff.

<!--
Visit this URL to learn more about how to write a pull request description:
https://armeria.dev/community/developer-guide#how-to-write-pull-request-description
-->

ikhoon commented on core/src/main/java/com/linecorp/armeria/client/retry/ExponentialBackoff.java:
```diff
@@ -95,4 +95,33 @@ public String toString() {
+        return new Builder();
+    }
+
+    public static class Builder {
```
Code style) We prefer top-level builder classes instead of nested classes.
Could you move this class to `ExpontionalBackoffBuilder`?

ikhoon commented on core/src/main/java/com/linecorp/armeria/client/retry/ExponentialBackoff.java:
```diff
@@ -95,4 +95,33 @@ public String toString() {
                           .toString();
     }
+
+    public static Builder builder() {
+        return new Builder();
+    }
```
Instead of adding a builder method to each implementation, should we add a builder method `Backoff` interface?
```java
interface Backoff {
   static ExpontionalBackoffBuilder builderForExpontional() {
      ...
   }
}
```
Note that `ExponentialBackoff` is a package-private class that users can't access.

ikhoon commented on core/src/main/java/com/linecorp/armeria/client/retry/ExponentialBackoff.java:
```diff
@@ -95,4 +95,33 @@ public String toString() {
+            return new ExponentialBackoff(initialDelayMillis, maxDelayMillis, multiplier);
+        }
+
+        public Builder initialDelayMillis(long initialDelayMillis) {
```
Should we make jitter values configured with this builder?

ikhoon commented on core/src/main/java/com/linecorp/armeria/client/retry/FibonacciBackoff.java:
```diff
@@ -88,4 +88,27 @@ public String toString() {
+        private long maxDelayMillis;
+
+        FibonacciBackoff build() {
+            return new FibonacciBackoff(initialDelayMillis, maxDelayMillis);
```
Should we migrate all factory methods where constructors are used into builder methods? e.g.
https://github.com/line/armeria/blob/ccb29fee0fea5a8657fe3ad2de1fe0a19062818c/core/src/main/java/com/linecorp/armeria/client/retry/Backoff.java#L71-L71

kth496 commented:
@ikhoon Thank you for your review. I will fix it within a week.

trustin commented:
@kth496 Looking forward to seeing more commits in this PR :heart_decoration:

trustin commented:
@kth496 Any updates?

kth496 commented:
@ikhoon @trustin 
Thanks for patience. I have improved the code based on your feedback:

- Moved the nested classes into top-level classes
- Moved the builder methods to the BackOff interface

Question:
Is there a good way to make the jitter value configurable in the builder? Since jitter can be commonly set across multiple BackOff implementations, I'm considering whether to define a separate higher-level builder.

ikhoon commented on core/src/main/java/com/linecorp/armeria/client/retry/ExponentialBackoffBuilder.java:
```diff
@@ -0,0 +1,98 @@
+    private long initialDelayMillis;
+    private long maxDelayMillis;
+    private double multiplier;
+
```
We prefer hiding the constructors of the builder class and guide to use `builder()` method.
```suggestion
    ExponentialBackoffBuilder() {}
```

ikhoon commented on core/src/main/java/com/linecorp/armeria/client/retry/ExponentialBackoffBuilder.java:
```diff
@@ -0,0 +1,98 @@
+ *
+ * @see ExponentialBackoff
+ */
+public class ExponentialBackoffBuilder {
```
Global comments:
- Add `@UnstableApi` to all new public classes.
- Add `final` keyword for builder classes.
```suggestion
@UnstableApi
public final class ExponentialBackoffBuilder {
```

ikhoon commented on core/src/main/java/com/linecorp/armeria/client/retry/ExponentialBackoffBuilder.java:
```diff
@@ -0,0 +1,98 @@
+     * @return this {@code ExponentialBackoffBuilder} instance for method chaining
+     */
+    public ExponentialBackoffBuilder initialDelayMillis(long initialDelayMillis) {
+        this.initialDelayMillis = initialDelayMillis;
```
Should we validate the input to ensure `initialDelayMillis >= 0`?
Please apply the same validation to other methods as well. 🙏

ikhoon commented on core/src/main/java/com/linecorp/armeria/client/retry/ExponentialBackoffBuilder.java:
```diff
@@ -0,0 +1,98 @@
+     *
+     * @return a newly created {@link ExponentialBackoff} with the configured delays and multiplier
+     */
+    ExponentialBackoff build() {
```
This must be `public`
```suggestion
    public ExponentialBackoff build() {
```

ikhoon commented on core/src/main/java/com/linecorp/armeria/client/retry/ExponentialBackoffBuilder.java:
```diff
@@ -0,0 +1,98 @@
+public class ExponentialBackoffBuilder {
+    private long initialDelayMillis;
+    private long maxDelayMillis;
+    private double multiplier;
```
Should we follow the default values used in the factory methods?
```suggestion
    private double multiplier = 2.0;
```

ikhoon commented on core/src/main/java/com/linecorp/armeria/client/retry/FibonacciBackoffBuilder.java:
```diff
@@ -0,0 +1,92 @@
+    }
+
+    /**
+     * Sets the maximum delay in milliseconds for the {@link FibonacciBackoff}.
```
`FibonacciBackoff` is an internal class to which we can't link in Javadoc.
```suggestion
     * Sets the maximum delay in milliseconds for the Fibonacci {@link Backoff}.
```

ikhoon commented on core/src/main/java/com/linecorp/armeria/client/retry/FixedBackoffBuilder.java:
```diff
@@ -0,0 +1,70 @@
+     *
+     * @return a newly created {@link FixedBackoff} with the configured delay
+     */
+    public Backoff build() {
```
Ditto. Please move to the bottom of this class.

ikhoon commented on core/src/main/java/com/linecorp/armeria/client/retry/FixedBackoffBuilder.java:
```diff
@@ -0,0 +1,70 @@
+    }
+
+    /**
+     * Sets the delay duration in milliseconds for the {@link FixedBackoff}.
```
```suggestion
     * Sets the delay duration in milliseconds for the fixed {@link Backoff}.
```

ikhoon commented on core/src/main/java/com/linecorp/armeria/client/retry/FixedBackoffBuilder.java:
```diff
@@ -0,0 +1,70 @@
+ *
+ * <pre>
+ * {@code
+ * FixedBackoff backoff = new FixedBackoffBuilder()
```
The constructor of `FixedBackoffBuilder` is now hidden.
```suggestion
 * FixedBackoff backoff = Backoff.builderForFixed()
```

ikhoon commented on core/src/main/java/com/linecorp/armeria/client/retry/FibonacciBackoffBuilder.java:
```diff
@@ -0,0 +1,92 @@
+    /**
+     * Builds and returns a new {@link FibonacciBackoff} instance with the configured
+     * initial and maximum delays.
+     *
+     * @return a newly created {@link FibonacciBackoff} with the configured delays
```
It seems redundant. 
```suggestion
```

ikhoon commented on core/src/main/java/com/linecorp/armeria/client/retry/FibonacciBackoffBuilder.java:
```diff
@@ -0,0 +1,92 @@
+ *
+ * <pre>
+ * {@code
+ * FibonacciBackoff backoff = new FibonacciBackoffBuilder()
```
Ditto. Please use builder methods.

ikhoon commented on core/src/main/java/com/linecorp/armeria/client/retry/ExponentialBackoffBuilder.java:
```diff
@@ -0,0 +1,108 @@
+ * @see ExponentialBackoff
+ */
+@UnstableApi
+public final class ExponentialBackoffBuilder {
```
It would be worth allowing to configure jitter and max attempts in all builder methods.
What do you think of introducing `AbstractBackoffBuilder` to set them?
```java
public abstract class AbstractBackoffBuilder<SELF extends AbstractBackoffBuilder<SELF>> {
    SELF jitter(double minJitterRate, double maxJitterRate, Supplier<Random> randomSupplier) {
        ... 
    }

    SELF maxAttempts(...) {
       ...
    }
}

public final class ExponentialBackoffBuilder 
    extends AbstractBackoffBuilder<ExponentialBackoffBuilder>{

    ...
}
```

kth496 commented on core/src/main/java/com/linecorp/armeria/client/retry/AbstractBackoffBuilder.java:
```diff
@@ -0,0 +1,83 @@
+        return self();
+    }
+
+    abstract Backoff doBuild();
+
+    /**
+     * Builds and returns {@link Backoff} instance with configured properties.
+     */
+    public final Backoff build() {
+        Backoff backoff = doBuild();
+        if (maxJitterRate > minJitterRate) {
+            backoff = new JitterAddingBackoff(backoff, minJitterRate, maxJitterRate, randomSupplier);
+        }
+        if (maxAttempts > 0) {
+            backoff = new AttemptLimitingBackoff(backoff, maxAttempts);
+        }
+        return backoff;
+    }
```
After much thought, I came up with the above solution. Is there a better way?

jrhee17 commented on core/src/main/java/com/linecorp/armeria/client/retry/AbstractBackoffBuilder.java:
```diff
@@ -0,0 +1,83 @@
+        return (SELF) this;
+    }
+
+    /**
+     * Sets the minimum and maximum jitter rates to apply to the delay.
+     */
+    SELF jitter(double minJitterRate, double maxJitterRate) {
+        this.minJitterRate = minJitterRate;
+        this.maxJitterRate = maxJitterRate;
+        return self();
+    }
+
+    /**
+     * Sets the minimum and maximum jitter rates to apply to the delay, as well as a
+     * custom {@link Random} supplier for generating the jitter.
+     */
+    SELF jitter(double minJitterRate, double maxJitterRate, Supplier<Random> randomSupplier) {
+        requireNonNull(randomSupplier, "randomSupplier");
+        this.minJitterRate = minJitterRate;
+        this.maxJitterRate = maxJitterRate;
+        this.randomSupplier = randomSupplier;
+        return self();
+    }
+
+    /**
+     * Sets the maximum number of attempts
+     * .
+     */
+    SELF maxAttempts(int maxAttempts) {
+        this.maxAttempts = maxAttempts;
+        return self();
+    }
```
Question) Shouldn't these methods be public?

jrhee17 commented on core/src/main/java/com/linecorp/armeria/client/retry/AbstractBackoffBuilder.java:
```diff
@@ -0,0 +1,83 @@
+     */
+    public final Backoff build() {
+        Backoff backoff = doBuild();
+        if (maxJitterRate > minJitterRate) {
+            backoff = new JitterAddingBackoff(backoff, minJitterRate, maxJitterRate, randomSupplier);
+        }
+        if (maxAttempts > 0) {
+            backoff = new AttemptLimitingBackoff(backoff, maxAttempts);
+        }
```
Can we just check nullability to decide whether to apply a wrapping backoff?

e.g.
```java
@Nullable
private Double minJitterRate;
...
if (minJitterRate != null && maxJitterRate != null) {
...
}
```

jrhee17 commented on core/src/main/java/com/linecorp/armeria/client/retry/AbstractBackoffBuilder.java:
```diff
@@ -0,0 +1,83 @@
+/**
+ * A skeletal builder implementation for {@link Backoff}.
+ */
+public abstract class AbstractBackoffBuilder<SELF extends AbstractBackoffBuilder<SELF>> {
```
Question) Does the abstract builder need to be public?

```suggestion
abstract class AbstractBackoffBuilder<SELF extends AbstractBackoffBuilder<SELF>> {
```

minwoox commented on core/src/main/java/com/linecorp/armeria/client/retry/ExponentialBackoffBuilder.java:
```diff
@@ -0,0 +1,101 @@
+ */
+@UnstableApi
+public final class ExponentialBackoffBuilder extends AbstractBackoffBuilder<ExponentialBackoffBuilder> {
+    private long initialDelayMillis;
+    private long maxDelayMillis;
```
Would you mind adding sensible default values for all the parameters in builders?

kth496 commented:
@minwoox 
Thank you for addressing the miscellaneous changes and committing them. I’ve also incorporated your additional review feedback!

minwoox commented on core/src/main/java/com/linecorp/armeria/client/retry/ExponentialBackoffBuilder.java:
```diff
@@ -0,0 +1,101 @@
+ */
+@UnstableApi
+public final class ExponentialBackoffBuilder extends AbstractBackoffBuilder<ExponentialBackoffBuilder> {
+    private long initialDelayMillis = 100;
```
Let's use 200:
https://github.com/line/armeria/blob/7c21957f42c1791e48c4980c814ac54047444a67/core/src/main/java/com/linecorp/armeria/common/DefaultFlagsProvider.java#L92

minwoox commented on core/src/main/java/com/linecorp/armeria/client/retry/FibonacciBackoffBuilder.java:
```diff
@@ -0,0 +1,85 @@
+ */
+@UnstableApi
+public final class FibonacciBackoffBuilder extends AbstractBackoffBuilder<FibonacciBackoffBuilder> {
+    private long initialDelayMillis;
+    private long maxDelayMillis;
```
Could you also add the default value for `FibonacciBackoffBuilder`, `FixedBackoffBuilder`, and `RandomBackoffBuilder`?

ikhoon reviewed (approved):
Thanks, @kth496! 👍🚀

minwoox reviewed (approved):
Nice work, @kth496! 👏 👏 👏
